exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/StatelessMarshalerUtil.java

```diff
@@ -0,0 +1,409 @@
+      Collection<T> dataList,
+      Function<T, Resource> getResource,
+      Function<T, InstrumentationScopeInfo> getInstrumentationScope,
+      MarshalerContext context) {
+    Map<Resource, Map<InstrumentationScopeInfo, List<T>>> result = context.getIdentityMap();
+
+    Grouper<T> grouper = context.getInstance(GROUPER_KEY, Grouper::new);
+    grouper.initialize(result, getResource, getInstrumentationScope, context);
+    dataList.forEach(grouper);
+
+    return result;
+  }
+
+  private static class Grouper<T> implements Consumer<T> {
+    @SuppressWarnings("NullAway")
+    private Map<Resource, Map<InstrumentationScopeInfo, List<T>>> result;
+
+    @SuppressWarnings("NullAway")
+    private Function<T, Resource> getResource;
+
+    @SuppressWarnings("NullAway")
+    private Function<T, InstrumentationScopeInfo> getInstrumentationScope;
+
+    @SuppressWarnings("NullAway")
+    private MarshalerContext context;
+
+    void initialize(
+        Map<Resource, Map<InstrumentationScopeInfo, List<T>>> result,
+        Function<T, Resource> getResource,
+        Function<T, InstrumentationScopeInfo> getInstrumentationScope,
+        MarshalerContext context) {
+      this.result = result;
+      this.getResource = getResource;
+      this.getInstrumentationScope = getInstrumentationScope;
+      this.context = context;
+    }
+
+    @Override
+    public void accept(T data) {
+      Resource resource = getResource.apply(data);
+      Map<InstrumentationScopeInfo, List<T>> scopeInfoListMap = result.get(resource);
+      if (scopeInfoListMap == null) {
+        scopeInfoListMap = context.getIdentityMap();
+        result.put(resource, scopeInfoListMap);
+      }
+      InstrumentationScopeInfo instrumentationScopeInfo = getInstrumentationScope.apply(data);
+      List<T> elementList = scopeInfoListMap.get(instrumentationScopeInfo);
+      if (elementList == null) {
+        elementList = context.getList();
+        scopeInfoListMap.put(instrumentationScopeInfo, elementList);
+      }
+      elementList.add(data);
+    }
+  }
+
+  /** Returns the size of a string field. */
+  public static int sizeString(
+      ProtoFieldInfo field, @Nullable String value, MarshalerContext context) {
+    if (value == null || value.isEmpty()) {
+      return sizeBytes(field, 0);
+    }
+    if (context.marshalStringNoAllocation()) {
+      int utf8Size = getUtf8Size(value, context);
+      context.addSize(utf8Size);
+      return sizeBytes(field, utf8Size);
+    } else {
+      byte[] valueUtf8 = MarshalerUtil.toBytes(value);
+      context.addData(valueUtf8);
+      return sizeBytes(field, valueUtf8.length);
+    }
+  }
+
+  /** Returns the size of a bytes field. */
+  public static int sizeBytes(ProtoFieldInfo field, int length) {
+    if (length == 0) {
+      return 0;
+    }
+    return field.getTagSize() + CodedOutputStream.computeLengthDelimitedFieldSize(length);
+  }
+
+  /** Returns the size of a repeated message field. */
```

Comment:
Can we note in the javadoc that that these methods change the state of marshaler context and need to be used in tandem with specific other methods to retrieve that state in the same order?